### PR TITLE
8332900: RISC-V: refactor nativeInst_riscv.cpp and macroAssembler_riscv.cpp

### PR DESCRIPTION
**Pull request by Hamlin-Li**
Hi,
Can you help to review the patch?
Currently, code in nativeInst_riscv.cpp and macroAssembler_riscv.cpp call each other, which is not right for readability and maintainance.
After refactoring, basically only code in nativeInst_riscv.cpp calls code in macroAssembler_riscv.cpp, but not in reverse direction.

Thanks!

* Tests are still running, so far so good.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332900](https://bugs.openjdk.org/browse/JDK-8332900): RISC-V: refactor nativeInst_riscv.cpp and macroAssembler_riscv.cpp (**Enhancement** - P4)

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**) ⚠️ Review applies to [ef11d6a7](https://git.openjdk.org/jdk/pull/19459/files/ef11d6a78a7d58016994bd5633ddbd76028e5013)
 * [Ludovic Henry](https://openjdk.org/census#luhenry) (@luhenry - Committer) ⚠️ Review applies to [ef11d6a7](https://git.openjdk.org/jdk/pull/19459/files/ef11d6a78a7d58016994bd5633ddbd76028e5013)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19459/head:pull/19459` \
`$ git checkout pull/19459`

Update a local copy of the PR: \
`$ git checkout pull/19459` \
`$ git pull https://git.openjdk.org/jdk.git pull/19459/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19459`

View PR using the GUI difftool: \
`$ git pr show -t 19459`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19459.diff">https://git.openjdk.org/jdk/pull/19459.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19459#issuecomment-2137751764)

### CODE REVIEW

**Review by RealFYang — COMMENTED** [review]
TBH, I don't think this is a change in the right direction. Classes like `NativeInstruction` and `MacroAssembler` are supposed to be there for different purposes, so I think it's not a issue for them to call each other where necessary. It's looks really strange to me to move functions like `extract_rs1` from `NativeInstruction` to `MacroAssembler`. It's a thing which is supposed to be in the domain of class `NativeInstruction`.

**Comment by Hamlin-Li** [comment]
I'm not sure, maybe because you're already familiar with current code, but I have a different opinion.

The only uses of `extract_rs1` are in static functions in `nativeInst_riscv.hpp`, so it's just a utility function, and does not matter where it's put.
But, e.g. things related to movptr[1/2] are currently separated in nativeInst_riscv and macroAssembler_riscv, which is not right, in one side it brings no benefit for reading the code, in the other side when you modify some code you need do code change at both files, https://github.com/openjdk/jdk/pull/19246 is an example.
The only things must be in nativeInst_riscv are the interfaces/fuctions it exposes to outside, but some implementation is not necessary to be in it, and it's better to be put in macroAssembler_riscv, as they're tightly coupled (e.g. movptr things).

**Comment by RealFYang** [comment]
I know what you mean. But they play different roles by design. Just think of `MacroAssembler` as producer and `NativeInstruction` as consumer/analyzer of the native instructions. There will surely be some protocols between the them under the hood, which I think is quite normal. BTW: this is also adopted by other CPUs like aarch64.

**Comment by Hamlin-Li** [comment]
> There will surely be some protocols between the them under the hood.

If this protocol means lots of dual direction communication, then we should consider if it's right (from a point of view OO design, it's an obvious code smell for me).
NativeInstruction could be a wrapper upon MacroAssembler, but not in reverse direction, in that way it makes things complicated and it's not necessary and bring no benefit.

> BTW: this is also adopted by other CPUs like aarch64.

Yes, we can refer or copy some code from other platforms in case they're well implemented.
Please check https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/nativeInst_aarch64.hpp#L137, the `extract` is not in nativeInst_aarch64, it's in assembler_aarch64.

**Comment by RealFYang** [comment]
> > There will surely be some protocols between the them under the hood.
> 
> If this protocol means lots of dual direction communication, then we should consider if it's right (from a point of view OO design, it's an obvious code smell for me). NativeInstruction could be a wrapper upon MacroAssembler, but not in reverse direction, in that way it makes things complicated and it's not necessary and bring no benefit.

I guess that might deserve a broader discussion as it does not seem to be a RISC-V specific issue.

> > BTW: this is also adopted by other CPUs like aarch64.
> 
> Yes, we can refer or copy some code from other platforms in case they're well implemented. Please check https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/nativeInst_aarch64.hpp#L137, the `extract` is not in nativeInst_aarch64, it's in assembler_aarch64.

The RISC-V counter part: https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/assembler_riscv.hpp#L384

The protocol is also there for aarch64: https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/nativeInst_aarch64.hpp#L290

**Comment by RealFYang** [comment]
> As I said, we can refer to other platforms in case they're well implemented. But, I see no benefit and reason to put e.g. `movptr1` and `patch_addr_in_movptr1` in macroAssembler and `is_movptr1_at` in nativeInst, I can not find the rational to do so, just because aarch64 do the similar thing? seems that's not a good reason; if it is, then how about other platforms, do they all follow aarch64 implementation details? Anyway, it's an implementation detail, we don't have to follow other platforms unless it's a good design choice for riscv too.

Apparently that's not what I mean. I just want say that it's a commonly adopted approach by other CPUs not just aarch64 and I am just fine with the current approach. But your opinion from a better OO design does make sense. Anyway, here might be a compromise for us.

I see you still kept following interfaces for `NativeInstruction`: 
```
  bool is_jal()                             const { return MacroAssembler::is_jal_at(addr_at(0));         }
  bool is_movptr()                          const { return MacroAssembler::is_movptr1_at(addr_at(0)) ||
                                                           MacroAssembler::is_movptr2_at(addr_at(0));     }
  bool is_movptr1()                         const { return MacroAssembler::is_movptr1_at(addr_at(0));     }
  bool is_movptr2()                         const { return MacroAssembler::is_movptr2_at(addr_at(0));     }
  bool is_auipc()                           const { return MacroAssembler::is_auipc_at(addr_at(0));       }
  bool is_call()                            const { return MacroAssembler::is_call_at(addr_at(0));        }
  bool is_jump()                            const { return MacroAssembler::is_jump_at(addr_at(0));        }
```

Can you do similar thing for other interfaces like `MacroAssembler::is_li16u_at`? Maybe `NativeInstruction::is_li16u` which delegates work to `MacroAssembler::is_li16u_at`. And we use these interfaces from `NativeInstruction::is_li16u` where appropriate at the callsites. Hopefully, I think the code will be more unified in style and resolves both of our concerns.

**Comment by Hamlin-Li** [comment]
> Like `NativeInstruction::is_li16u` which delegates work to `MacroAssembler::is_li16u_at`.

I don't find `NativeInstruction::is_li16u`, maybe you want to say something else for the delegation you mentioned?

Take `MacroAssembler::is_li16u_at` as example, I moved it to macroAssembler, because in macroAssembler it's used too. So one of the principals I'd like to stick to in this refactoring is to make these 2 classes's communication unidirectional, so maybe it's better to move `MacroAssembler::is_li16u_at` too.

**Comment by RealFYang** [comment]
> > Like `NativeInstruction::is_li16u` which delegates work to `MacroAssembler::is_li16u_at`.
> 
> I don't find `NativeInstruction::is_li16u`, maybe you want to say something else for the delegation you mentioned?

Never mind. I think I miss read the code.

> Take `MacroAssembler::is_li16u_at` as example, I moved it to macroAssembler, because in macroAssembler it's used too. So one of the principals I'd like to stick to in this refactoring is to make these 2 classes's communication unidirectional, so maybe it's better to move `MacroAssembler::is_li16u_at` too.

Yeah. Your changes now becomes interesting to me now. I am having another check.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp [hunk 1]
```diff
@@ -41,15 +40,244 @@
...
+  }
+
+  static bool is_call_at(address instr) {
+    if (is_jal_at(instr) || is_jalr_at(instr)) {
+      return true;
+    }
+    return false;
+  }
+
+ private:
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:99]: Maybe it's better to make those is_XXX functions public?

**Review by RealFYang — COMMENTED** [review]
LGTM modulo one minor comment. I think you are right in this refactoring work. Sorry for misunderstanding your opinion. Thanks!

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp (hunk 1 above)
Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.hpp]: Yes, modified.

**Comment by Hamlin-Li** [comment]
> LGTM modulo one minor comment. I think you are right in this refactoring work. Sorry for mis-understanding your opinion. Thanks!

No worry. Thanks for discussion!

**Review by RealFYang — APPROVED** [review]

**Comment by Hamlin-Li** [comment]
> Hi Halim, this will cause merge hell for me in #19453.
> 
> You have a bunch of stuff in hpp files which can be moved to cpp file, to keep headers clean and not export so. E.g. check_movptr1_data_dependency is only used in macroAssembler_riscv.cpp, now everyone including MASM.hpp get copy of this method (which is compiled away). I.e. if you want it to be a member method move the definition to cpp and just keep the declaration in hpp.

I'm not sure.
If this is a problem in this patch, then in original implementation, MASM.hpp includes nativeInst.hpp, it will cause the similar issue?
In another hand, will unused methods in a header file still exist in the compiled binary?

The reason I do it this way is because I just want to refactor necessary things (the pricipal is to only nativeInst depends on macroAssembler, not in reverse direction), for other things, I try to keep it as original ones. So if there is other optimization opportunities, it's better to do it in separate pr's.

**Review by luhenry — APPROVED** [review]

**Comment by robehn** [comment]
> I'm not sure. If this is a problem in this patch, then in original implementation, MASM.hpp includes nativeInst.hpp, it will cause the similar issue? In another hand, will unused methods in a header file still exist in the compiled binary?

It's about having readable header, havin long private, not part of API, method in header clobber it.
They will not exists, but compiling away methods is not free as it prolongs the compile time.

> 
> The reason I do it this way is because I just want to refactor necessary things (the pricipal is to only nativeInst depends on macroAssembler, not in reverse direction), for other things, I try to keep it as original ones. So if there is other optimization opportunities, it's better to do it in separate pr's.

If this is a simple move of code, then I suggest you wait until my patch, as that changes names and doubles the methods which you want to move.

Also now NativeCall::instruction_size, which is used in shared code is nativeInst, but NativeInstruction::instruction_size is in MASM. From my patch riscv.ad file:

```
int MachCallDynamicJavaNode::ret_addr_offset()
{
  return NativeMovConstReg::movptr2_instruction_size +  NativeInstruction::instruction_size; // movptr2, jal
  if (UseTrampolines) {
    return NativeMovConstReg::movptr2_instruction_size +  NativeInstruction::instruction_size; // movptr2, jal
  }
}
```

**Comment by robehn** [comment]
```
class Assembler : public AbstractAssembler {
public:

  enum {
    instruction_size = 4,

class MacroAssembler: public Assembler {

 public:
  enum {
    instruction_size = 4,
```

Not sure why you have the same enum defs in both asm and masm?

**Comment by Hamlin-Li** [comment]
> > I'm not sure. If this is a problem in this patch, then in original implementation, MASM.hpp includes nativeInst.hpp, it will cause the similar issue? In another hand, will unused methods in a header file still exist in the compiled binary?
> 
> It's about having readable header, havin long private, not part of API, method in header clobber it.

Not sure, as it's in its own section, not mixed with other sections.
But, it's fine, I'll move it to the bottom of the file.

> They will not exists, but compiling away methods is not free as it prolongs the compile time.
As I said, MASM.hpp included nativeInst.hpp before, so it's not a issue (or new issue)?

> 
> > The reason I do it this way is because I just want to refactor necessary things (the pricipal is to only nativeInst depends on macroAssembler, not in reverse direction), for other things, I try to keep it as original ones. So if there is other optimization opportunities, it's better to do it in separate pr's.
> 
> If this is a simple move of code, then I suggest you wait until my patch, as that changes names and doubles the methods which you want to move.

It's just a merge, either in this pr or that pr, not that complicated?

> 
> Also now NativeCall::instruction_size, which is used in shared code is nativeInst, but NativeInstruction::instruction_size is (duplicated) in MASM. EDIT, better example:
> 
> ```
> CodeBuffer* PhaseOutput::init_buffer() {
> ...
>   int pad_req   = NativeCall::instruction_size;
> ```
> 
> So there is duplication if all size both in nativeInst and masm.

I don't get the point, can you clarify?

**Comment by Hamlin-Li** [comment]
> ```
> class Assembler : public AbstractAssembler {
> public:
> 
>   enum {
>     instruction_size = 4,
> 
> class MacroAssembler: public Assembler {
> 
>  public:
>   enum {
>     instruction_size = 4,
> ```
> 
> Not sure why you have the same enum defs in both asm and masm?



> ```
> class Assembler : public AbstractAssembler {
> public:
> 
>   enum {
>     instruction_size = 4,
> 
> class MacroAssembler: public Assembler {
> 
>  public:
>   enum {
>     instruction_size = 4,
> ```
> 
> Not sure why you have the same enum defs in both asm and masm?

Yeh, it's redundant, I'll remove it. Thanks.

**Comment by robehn** [comment]
> I don't get the point, can you clarify?

I think don't de-coupling by duplicating the constant is better than keeping the depedancy.
As long as the shared code expects the NativeXXX::instruction_size to work we can't move them.

E.g. (partly from my patch) nativeInst.cpp
```
if (NativeShortCall::is_at(addr)) {
   return NativeShortCall::instruction_size;
} else if (NativeFarCall::is_at(addr)) {
   return NativeFarCall::instruction_size;
}
```

In masm related code:
```
int MacroAssembler::max_patchable_far_call_stub_size() {                          
  // Max stub size: alignment nop, TrampolineStub.
  if (UseTrampolines) {
    return NativeInstruction::instruction_size + NativeShortCall::trampoline_size;
```

This would now look like:
```
int MacroAssembler::max_patchable_far_call_stub_size() {                          
  // Max stub size: alignment nop, TrampolineStub.
  if (UseTrampolines) {
    return Masm::instruction_size + Masm::native_short_call_trampoline_size;
```

Using different constant here is not an improvement.

**Comment by Hamlin-Li** [comment]
> > I don't get the point, can you clarify?
> 
> I think don't de-coupling by duplicating the constant is better than keeping the depedancy. As long as the shared code expects the NativeXXX::instruction_size to work we can't move them.
> 
> E.g. (partly from my patch) nativeInst.cpp
> 
> ```
> if (NativeShortCall::is_at(addr)) {
>    return NativeShortCall::instruction_size;
> } else if (NativeFarCall::is_at(addr)) {
>    return NativeFarCall::instruction_size;
> }
> ```
> 
> In masm related code:
> 
> ```
> int MacroAssembler::max_patchable_far_call_stub_size() {                          
>   // Max stub size: alignment nop, TrampolineStub.
>   if (UseTrampolines) {
>     return NativeInstruction::instruction_size + NativeShortCall::trampoline_size;
> ```
> 
> This would now look like:
> 
> ```
> int MacroAssembler::max_patchable_far_call_stub_size() {                          
>   // Max stub size: alignment nop, TrampolineStub.
>   if (UseTrampolines) {
>     return Masm::instruction_size + Masm::native_short_call_trampoline_size;
> ```
> 
> Using different constant here is not an improvement.

What I want is to let MASM have no dependency on nativeInst, that's one of the principals I stick to in this pr, as dependency on nativeInst from MASM makes no sense from the point of view of code logic, and bidirectional dependency makes code dependency complicated which will also lead to issues of readability and maintainance. And if you are going to do it, I hope to do it thoroughly.

**Comment by robehn** [comment]
> What I want is to let MASM have no dependency on nativeInst, that's one of the principals I stick to in this pr, as dependency on nativeInst from MASM makes no sense from the point of view of code logic, and bidirectional dependency makes code dependency complicated which will also lead to issues of readability and maintainance. And if you are going to do it, I hope to do it thoroughly.

I'm not disagreeing with your intent.
I'm just saying I duplicating the constants is not good solution, and the code in some cases will be worse.

**Comment by Hamlin-Li** [comment]
> I'm not disagreeing with your intent. I'm just saying I duplicating the constants is not good solution, and the code in some cases will be worse.

Thanks for clarifying, I get your point.
But still think we'd better have no dependency from macroAssembler to nativeInst. And I don't see issue in following code (in your example above)
```
int MacroAssembler::max_patchable_far_call_stub_size() {                          
  // Max stub size: alignment nop, TrampolineStub.
  if (UseTrampolines) {
    return Masm::instruction_size + Masm::native_short_call_trampoline_size;
```
`MacroAssembler` should only know fields or methods in itself, not in nativeInst. 
If some code need to access both `MacroAssembler` and `nativeInst`, it should know both of them.
And, if must choose between bidirectional dependencies and some data duplication, I prefer the latter.

And in existing code of nativeInst, there are some kind of similar data duplication codes, e.g. : https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/nativeInst_riscv.hpp#L446

**Comment by robehn** [comment]
Yea, there is were we disagree, as constants needs to be duplicate there is a depenecy even if you now can remove header.
If guys like, sure. If it is possible for you to wait for the other PR, please do so.

**Comment by Hamlin-Li** [comment]
Thanks @RealFYang @luhenry @robehn for your reviewing.

/integrate

**Pull request closed without merging**